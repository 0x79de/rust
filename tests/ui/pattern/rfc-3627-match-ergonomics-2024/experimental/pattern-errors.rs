//@ edition: 2024
//@ revisions: classic2024 structural2024
//! Test cases for poorly-typed patterns in edition 2024 which are caught by HIR typeck. These must
//! be separate from cases caught by MIR borrowck or the latter errors may not be emitted.
#![allow(incomplete_features)]
#![cfg_attr(classic2024, feature(ref_pat_eat_one_layer_2024))]
#![cfg_attr(structural2024, feature(ref_pat_eat_one_layer_2024_structural))]

pub fn main() {
    if let Some(&mut x) = &Some(&mut 0) {
        //[classic2024]~^ ERROR: mismatched types
        let _: &u32 = x;
    }
    if let Some(&mut Some(&x)) = &Some(&mut Some(0)) {
        //[classic2024]~^ ERROR: mismatched types
        let _: u32 = x;
    }
    if let Some(Some(&mut x)) = &Some(Some(&mut 0)) {
        //[classic2024]~^ ERROR: mismatched types
        let _: &u32 = x;
    }

    if let Some(&mut Some(&_)) = &Some(&Some(0)) {
        //~^ ERROR: mismatched types
    }
    if let Some(&Some(&mut _)) = &Some(&mut Some(0)) {
        //[structural2024]~^ ERROR: mismatched types
    }
    if let Some(&Some(&mut _)) = &mut Some(&Some(0)) {
        //~^ ERROR: mismatched types
    }
    if let Some(&Some(Some(&mut _))) = &Some(Some(&mut Some(0))) {
        //[structural2024]~^ ERROR: mismatched types
    }
    if let Some(&mut Some(x)) = &Some(Some(0)) {
        //~^ ERROR: mismatched types
    }
}

fn structural_errors_0() {
    let &[&mut x] = &&mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let &[&mut x] = &mut &mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let &[&mut ref x] = &&mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let &[&mut ref x] = &mut &mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let &[&mut mut x] = &&mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let &[&mut mut x] = &mut &mut [0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;
}

fn structural_errors_1() {
    let [&(mut x)] = &[&0];
    //[structural2024]~^ ERROR: binding cannot be both mutable and by-reference
    let _: &u32 = x;

    let [&(mut x)] = &mut [&0];
    //[structural2024]~^ ERROR: binding cannot be both mutable and by-reference
    let _: &u32 = x;
}

fn structural_errors_2() {
    let [&&mut x] = &[&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let [&&mut x] = &mut [&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let [&&mut ref x] = &[&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let [&&mut ref x] = &mut [&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let [&&mut mut x] = &[&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let [&&mut mut x] = &mut [&mut 0];
    //[structural2024]~^ ERROR: mismatched types
    //[structural2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;
}

fn classic_errors_0() {
    let [&mut x] = &[&mut 0];
    //[classic2024]~^ ERROR: mismatched types
    //[classic2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let [&mut &x] = &[&mut 0];
    //[classic2024]~^ ERROR: mismatched types
    //[classic2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;

    let [&mut &ref x] = &[&mut 0];
    //[classic2024]~^ ERROR: mismatched types
    //[classic2024]~| cannot match inherited `&` with `&mut` pattern
    let _: &u32 = x;

    let [&mut &(mut x)] = &[&mut 0];
    //[classic2024]~^ ERROR: mismatched types
    //[classic2024]~| cannot match inherited `&` with `&mut` pattern
    let _: u32 = x;
}
