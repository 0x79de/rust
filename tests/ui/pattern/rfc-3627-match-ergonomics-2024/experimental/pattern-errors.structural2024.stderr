error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:23:17
   |
LL |     if let Some(&mut Some(&_)) = &Some(&Some(0)) {
   |                 ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     if let Some(&Some(&_)) = &Some(&Some(0)) {
   |                 ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:26:23
   |
LL |     if let Some(&Some(&mut _)) = &Some(&mut Some(0)) {
   |                       ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     if let Some(&Some(&_)) = &Some(&mut Some(0)) {
   |                       ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:29:23
   |
LL |     if let Some(&Some(&mut _)) = &mut Some(&Some(0)) {
   |                       ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     if let Some(&Some(&_)) = &mut Some(&Some(0)) {
   |                       ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:32:28
   |
LL |     if let Some(&Some(Some(&mut _))) = &Some(Some(&mut Some(0))) {
   |                            ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     if let Some(&Some(Some(&_))) = &Some(Some(&mut Some(0))) {
   |                            ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:35:17
   |
LL |     if let Some(&mut Some(x)) = &Some(Some(0)) {
   |                 ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     if let Some(&Some(x)) = &Some(Some(0)) {
   |                 ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:41:11
   |
LL |     let &[&mut x] = &&mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&x] = &&mut [0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:46:11
   |
LL |     let &[&mut x] = &mut &mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&x] = &mut &mut [0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:51:11
   |
LL |     let &[&mut ref x] = &&mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&ref x] = &&mut [0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:56:11
   |
LL |     let &[&mut ref x] = &mut &mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&ref x] = &mut &mut [0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:61:11
   |
LL |     let &[&mut mut x] = &&mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&mut x] = &&mut [0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:66:11
   |
LL |     let &[&mut mut x] = &mut &mut [0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let &[&mut x] = &mut &mut [0];
   |           ~

error[E0658]: binding cannot be both mutable and by-reference
  --> $DIR/pattern-errors.rs:73:12
   |
LL |     let [&(mut x)] = &[&0];
   |            ^^^^
   |
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = help: add `#![feature(mut_ref)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: binding cannot be both mutable and by-reference
  --> $DIR/pattern-errors.rs:77:12
   |
LL |     let [&(mut x)] = &mut [&0];
   |            ^^^^
   |
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = help: add `#![feature(mut_ref)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:83:11
   |
LL |     let [&&mut x] = &[&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&x] = &[&mut 0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:88:11
   |
LL |     let [&&mut x] = &mut [&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&x] = &mut [&mut 0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:93:11
   |
LL |     let [&&mut ref x] = &[&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&ref x] = &[&mut 0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:98:11
   |
LL |     let [&&mut ref x] = &mut [&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&ref x] = &mut [&mut 0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:103:11
   |
LL |     let [&&mut mut x] = &[&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&mut x] = &[&mut 0];
   |           ~

error[E0308]: mismatched types
  --> $DIR/pattern-errors.rs:108:11
   |
LL |     let [&&mut mut x] = &mut [&mut 0];
   |           ^^^^^
   |
   = note: cannot match inherited `&` with `&mut` pattern
help: replace this `&mut` pattern with `&`
   |
LL |     let [&&mut x] = &mut [&mut 0];
   |           ~

error: aborting due to 19 previous errors

Some errors have detailed explanations: E0308, E0658.
For more information about an error, try `rustc --explain E0308`.
